tests-arquillian/src/test/java/org/jboss/weld/tests/interceptors/producer/InterceptionFactoryTest.java

```diff
@@ -138,27 +138,55 @@ public void testFactoryFromGenericInterface(@Produced SomeGenericInterface<List<
         assertEquals(2, Producer.INVOCATIONS.size());
     }
 
-    @Inject
-    @Any
-    Instance<Object> beanInstance;
+    @Test
+    public void testFactoryBackedByUnproxyableImpl(@Produced ProxyableInterface bean) {
+        // this test adds class level Monitor and method level Hello bindings
+        Producer.reset();
+        assertEquals("Hello " + UnproxyableImpl.class.getSimpleName(), bean.ping());
+        assertEquals(1, Producer.INVOCATIONS.size());
+    }
+
+    @Test
+    public void testFactoryFromInterfaceWithMethodAnnotation(@Produced ProxyableInterfaceWithMethodAnnotation bean) {
+        // this test adds class level Monitor and method level Hello bindings
+        Producer.reset();
+        assertEquals("Hello " + UnproxyableInterfaceWithMethodAnnotationImpl.class.getSimpleName(), bean.ping());
+        assertEquals(1, Producer.INVOCATIONS.size());
+    }
 
     @Test
-    public void testFactoryFromInterfaceWithMethodAnnotation() {
-        try {
-            beanInstance.select(UninterestingInterface1.class, Produced.Literal.INSTANCE).get().ping();
-            Assert.fail();
-        } catch (IllegalStateException e) {
-            // expected
-        }
+    public void testFactoryFromInterfaceWithClassAnnotation(@Produced ProxyableInterfaceWithClassAnnotation bean) {
+        // this test adds class level Monitor and method level Hello bindings
+        Producer.reset();
+        assertEquals("Hello " + UnproxyableInterfaceWithClassAnnotationImpl.class.getSimpleName(), bean.ping());
+        assertEquals(1, Producer.INVOCATIONS.size());
     }
 
     @Test
-    public void testFactoryFromInterfaceWithTypeAnnotation() {
-        try {
-            beanInstance.select(UninterestingInterface2.class, Produced.Literal.INSTANCE).get().ping();
-            Assert.fail();
-        } catch (IllegalStateException e) {
-            // expected
-        }
+    public void testFactoryFromGenericInterfaceStructure(@Produced InterfaceWithGenericsB<String, Integer> bean) {
+        Producer.reset();
+        bean.ping(); // this doesn't trigger anything as we don't inherit annotations ATM
+        assertEquals("Hello " + UnproxyableInterfaceWithGenericsChainImpl.class.getSimpleName(), bean.pong());
+        assertEquals(1, Producer.INVOCATIONS.size());
+    }
+
+    @Test
+    public void testFactoryWhenInterfaceHasEqualAnnotationsToThoseWeAreAdding(@Produced InterfaceWithAnnotation bean) {
+        // Currently, we do not verify this shouldn't happen, checking it would require comparison between
+        // added annotation and the AT we create from interface
+        // hence this test should just "silently succeed"
```

Comment:
Another grey area, do we want to resolve clashes when there is a binding on interface and we add one? I am not sure we should as normally you need to place interceptor bindings on bean class anyway.

Reply:
I don't think we should resolve the clashes. We don't do this for `ProcessAnnotatedType` and friends either.